### PR TITLE
Revamp the `Layouts` page

### LINKED ISSUE #2527: Revamp the `Layouts` page
As agreed in [the site structure document](https://docs.google.com/document/d/10Fu7oqDzdM_D6LbexzwX9arh51Tic7AGvkWTQrL6jjQ/edit), revamp the _"Layouts"_ page such that

- Outline:
  - What is a layout?
    - List layouts and link to the associated plugin reference page
  - Layouts with dedicated pages
    - Pattern Layout
      - Fix #2518 
    - JSON Template Layout
  - Extending layouts
    - Difference between `Encoder` and `Layout`
    - Note about garbage-free methods
- Address #1950
- Target `2.x`, port to `main`

### PR DESCRIPTION
**Pull request by vy**
Fixes #2527.

### CODE REVIEW

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc [hunk 1]
```diff
@@ -0,0 +1,1719 @@
...
+----
+
+The call site can look like this:
+
+[source,text]
+----
+logger.info("@\|KeyStyle {}\|@ = @\|ValueStyle {}\|@", entry.getKey(), entry.getValue());
+----
+
+You can enable xref:manual/lookups.adoc[] using `\{lookups}` and log messages like `${date:YYYY-MM-dd}` using lookups.
```
Comment (ppkarwasz) [inline src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:1097]: Can you really evaluate _lookups_ in **messages**?
I think this functionality was removed around `2.17.x`.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc [hunk 2]
```diff
@@ -0,0 +1,1719 @@
...
+==== Highlight
+
+Adds ANSI colors to the result of the enclosed pattern based on the current event's xref:manual/customloglevels.adoc[logging level].
+Windows users should refer to xref:#jansi[].
+
+.link:../javadoc/log4j-core/org/apache/logging/log4j/core/pattern/HighlightConverter.html[`HighlightConverter`] specifier grammar
+[source,text]
+----
+highlight{pattern}{style}
+----
```
Comment (ppkarwasz) [inline src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:698]: The `style` attribute is by far the most complex one.
Its syntax is a comma-separated list of specifiers in the form:

- `DISABLEANSI`, `NOCONSOLENOANSI`: we might want to mention that these specifiers are accepted for BC, but redirect users to the `disableAnsi` and `noConsoleNoAnsi` configuration attributes of `PatternLayout`.
- `STYLE=(DEFAULT|LOGBACK)`: is replaced by a predefined map level->color specifier. This **should** appear first in the list.
- `<level_name>=<color_specifier>`: the `<color_specifier>` part is again a complex structure. We should consider creating a separate section on how to specify colors, since the logic is shared with `%style`.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc [hunk 3]
```diff
@@ -0,0 +1,1719 @@
...
+Use ANSI escape sequences to style the result of the enclosed pattern.
+Windows users should refer to xref:#jansi[].
+
+.link:../javadoc/log4j-core/org/apache/logging/log4j/core/pattern/StyleConverter.html[`StyleConverter`] specifier grammar
+[source,text]
+----
+style{pattern}{ANSI style}
+----
+
+The style can consist of a comma-separated list of style names from the following table:
```
Comment (ppkarwasz) [inline src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:1273]: For compatibility with `%highlight` this can also be a space-separated list.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc [hunk 4]
```diff
@@ -0,0 +1,1719 @@
...
+
+.link:../javadoc/log4j-core/org/apache/logging/log4j/core/pattern/UuidConverter.html[`UuidConverter`] specifier grammar
+[source,text]
+----
+u{RANDOM|TIME}
+uuid{RANDOM|TIME}
+----
+
+The time-based UUID is a Type 1 UUID generated using the MAC address of each host
+To ensure uniqueness across multiple JVMs and/or class loaders on the same host, a random number between 0 and 16,384 will be associated with each instance of the UUID generator class, and included in each time-based UUID generated.
```
Comment (ppkarwasz) [inline src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:1461]:
```suggestion
To ensure uniqueness across multiple JVMs and/or class loaders on the same host, a random number between 0 and 16,384 will be associated with each instance of the UUID generator class, and included in each time-based UUID generated.
See also xref:manual/systemproperties.adoc#log4j2.uuidSequence[`log4j2.uuidSequence`].
```

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc [hunk 5]
```diff
@@ -0,0 +1,1719 @@
...
+
+|xref:#converter-map[%K\{key} +
+%map\{key} +
+%MAP\{key}]
+|
+
+|xref:#converter-message[%m +
+%msg +
+%message]
+|Garbage-free unless message text contains `${`
```
Comment (ppkarwasz) [inline src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:1598]: I think this does not apply any more.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc [hunk 6]
```diff
@@ -0,0 +1,1719 @@
...
+|xref:#converter-thread-priority[%tp]
+|
+
+|xref:#converter-thread-context-map[%X{key[,key2...\]} +
+%mdc{key[,key2...\]} +
+%MDC{key[,key2...\]}]
+|
+
+|literal text
+|Garbage-free unless it contains `${`
```
Comment (ppkarwasz) [inline src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:1657]: We should probably add a note about lazy substitution.

Adding `${k8s:containerId}` is garbage-free, because it will be expanded at configuration time. Adding `$${date:hh:mm:ss.SSS}` is not, since it will be expanded at each log event.

**Review by ppkarwasz — CHANGES_REQUESTED** [review]
Nice job!

I think we should profit from this opportunity and fix some problems we have in the documentation of layouts (e.g. `PatternLayout`), as in the remarks below.

We could also include a link to the plugin reference, whenever we deal with a plugin, so that users can find the errors in the documentation themselves.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc [hunk 7]
```diff
@@ -0,0 +1,1719 @@
...
+include::example$manual/pattern-layout/script-pattern-selector/log4j2.properties[lines=19..38,indent=0]
+----
+====
+
+[#converters]
+=== Pattern converters
+
+The Pattern Layout _conversion pattern_ is composed of literal text and format control expressions called _conversion specifiers_ – refer to xref:#usage[] for details.
+xref:manual/plugins.adoc[Plugins] implementing link:../javadoc/log4j-core/org/apache/logging/log4j/core/pattern/PatternConverter.html[`PatternConverter`] are admitted to the *pattern converter* registry of Pattern Layout, and used to resolve the conversion specifiers.
+The predefined set of pattern converters are shared below.
```
[the remark this answers is not in the document]
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc]: Fixed in b95d668f7e9dfe7a7d5b7787be58f4a7b9b65a15.

File: src/site/antora/modules/ROOT/pages/manual/layouts.adoc [hunk 8]
```diff
@@ -14,2120 +14,863 @@
...
+[#concerns]
 == Common concerns
 
+This section introduces you to some common concerns shared by almost all xref:#collection[predefined layouts] that you need to be aware of while using them.
+
+[#charset]
+=== Character encoding
+
+All xref:#collection[predefined layouts] produce `String` that eventually get converted into a `byte[]` using the https://docs.oracle.com/javase/8/docs/api/java/nio/charset/Charset.html[`Charset`] configured.
+While doing so, they use `UTF-8` by default.
```
[the remark this answers is not in the document]
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/layouts.adoc]: Except `SerializedLayout` and `JsonTemplateLayout`, all layouts extend from `AbstractStringLayout`, which defaults the `charset` to `UTF-8`. `PatternLayout` deviates from this and uses the platform default due to [LOG4J2-783](https://issues.apache.org/jira/browse/LOG4J2-783). I will rephrase this as follows:

> While doing so, unless an explicit encoding configuration is stated, they use `UTF-8` by default.

Fixed in 74a6ae462b201808bc06ca86888fc589b3adccd4.

For the record, `pattern-layout.adoc` was explicitly stating that `charset` defaults to the platform default.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc (hunk 2 above)
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:719]: @ppkarwasz, would you mind pushing this improvement in a commit to this branch, please?

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc [hunk 9]
```diff
@@ -0,0 +1,1719 @@
...
+
+A logger conversion specifier can be optionally followed by a _precision specifier_, which consists of a decimal integer, or a pattern starting with a decimal integer.
+
+* When the precision specifier is an integer value, it reduces the size of the logger name.
+If the number is positive, the layout prints the corresponding number of the rightmost logger name components.
+If negative, the layout removes the corresponding number of leftmost logger name components.
+* If the precision contains periods then the number before the first period identifies the length to be printed from items that precede tokens in the rest of the pattern.
+If the number after the first period is followed by an asterisk it indicates how many of the rightmost tokens will be printed in full.
+* If the precision contains any non-integer characters, then the layout abbreviates the name based on the pattern.
+If the precision integer is less than one, the layout still prints the right-most token in full.
```
[the remark this answers is not in the document]
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:947]: Totally agreed that the current description is extremely confusing. Would you mind pushing the improvement you stated in a commit to this branch, please?

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc (hunk 1 above)
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc]: Fixed in 76f77b59edc56a27ae277dbcd8c605c0a04841ff.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc (hunk 3 above)
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:1283]: I'd appreciate it if you can address this while improving the section on `%highlight`.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc (hunk 6 above)
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc]: Fixed in f77b5c0fce41b93da448f5d78fe902a09f970092.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc (hunk 5 above)
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc]: Fixed in f77b5c0fce41b93da448f5d78fe902a09f970092.

File: src/site/antora/modules/ROOT/pages/manual/json-template-layout.adoc [hunk 10]
```diff
@@ -1983,10 +2046,10 @@ To get the most out of it, mind the following checklist:
 [#faq-lookups]
 === Are lookups supported in templates?
 
-Yes, xref:manual/lookups.adoc[lookups] (e.g., `${java:version}`,
-`${env:USER}`, `${date:MM-dd-yyyy}`) are supported in string
-literals of templates. Though note that they are not garbage-free.
+Yes, xref:manual/lookups.adoc[lookups] (e.g., `${java:version}`, `${env:USER}`, `${date:MM-dd-yyyy}`) are supported in string literals of templates.
+Though note that they are not garbage-free.
```
[the remark this answers is not in the document]
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/json-template-layout.adoc]: Fixed in c5f012682d393ddb372a8b3cb0d2d184440a30f0.

File: src/site/antora/modules/ROOT/pages/manual/json-template-layout.adoc [hunk 11]
```diff
@@ -16,422 +16,479 @@
...
-| recycling strategy that can either be `dummy`, `threadLocal`, or `queue`
-  (set by `log4j.layout.jsonTemplate.recyclerFactory` property)
+|Type |`Charset`
+|Default value |`UTF-8`
+|Configuration property |`log4j.layout.jsonTemplate.charset`
 |===
 
-[#additional-event-template-fields]
-=== Additional event template fields
+`Charset` used for encoding the produced JSON into bytes
```
[the remark this answers is not in the document]
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/json-template-layout.adoc:199]: Fixed in c5f012682d393ddb372a8b3cb0d2d184440a30f0.

File: src/site/antora/modules/ROOT/pages/manual/json-template-layout.adoc [hunk 12]
```diff
@@ -16,422 +16,479 @@
...
+[cols="2h,6"]
+|===
+|Type |`boolean`
+|Default value |`false`
+|Configuration property |`log4j.layout.jsonTemplate.locationInfoEnabled`
+|===
 
-The default `format` for the added new fields are `String`.
-One can also provide JSON-formatted additional fields:
+Toggles access to the `LogEvent` source; file name, line number, etc.
```
[the remark this answers is not in the document]
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/json-template-layout.adoc:216]: Fixed in c5f012682d393ddb372a8b3cb0d2d184440a30f0.

**Comment by vy** [comment]
> We could also include a link to the plugin reference, whenever we deal with a plugin, so that users can find the errors in the documentation themselves.

@ppkarwasz, added in 0cf6c346e775b6307a3b0d7f793835c94f6f2761.

File: src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc (hunk 2 above)
Comment (ppkarwasz) [reply src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc:721]: Fixed in 8950932be4e2865d5fc2cddce84da2b4cb52e940, except for the `DISABLEANSI` and `NOCONSOLEANSI`, which seem more like an internal hack to pass the value of the `PatternLayout` configuration options to the converter, than real options.
